### ISSUE TITLE
MD005 False Positive When Both Ordered and Unordered Lists are Used

### ISSUE THREAD

**Issue by miralexan** [markdownlint/markdownlint#296]
The following markdown snippet triggers an MD005 violation:

```markdown
1. OL level one
   1. OL level two

* UL level one
  * UL level two
```

Output:

```
test.md:5: MDO05 Inconsistent indentation for list items at the same level

A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md
```

It seems that when comparing indentation levels, mdl doesn't distinguish between ordered and unordered lists.

**Comment by jaymzh** [comment]
This seems desired to me... the same number of spaces should be used (preferably, IMO, 2, always).

**Comment by miralexan** [comment]
@jaymzh The markdown style I'm most familiar with says that

* Nested lists should be indented so that the nested list's start token lines up with the parent's text.
* There should be a space after a list's start token.

So two spaces for each level of an unordered list and three spaces for each level of an ordered list.

***

Also, for my own reference, when you say that 2 spaces should be preferred, what does that look like with ordered lists? Like this?

```markdown
1. First item
  1. Second item
```

**Comment by patrickhousley** [comment]
The below markdown, when linted, violates rule MD005 for the nested unordered list part. I have tried 3 and 4 spaces. The only way to make it work is to de-indent the unordered list but then it is no longer nested inside the ordered list.

```markdown
1. Select the Group or project to create the board.
1. Select "Issues" from Left Column menu
1. Select "Board" from Issue Menu list
1. Open Dropdown for Switch board at top of page.
    *The default is Developement*
1. Select "Create New Board"
1. Enter Board name
1. Expand Board scope.
1. Enter desired scope.
    - Milestone
    - Labels
    - Assignees
    - Weights
1. Click "Create Board"
```

```bash
09:59 $ mdl content
content/docs/framework/Agile_Documentation.md:87: MD005 Inconsistent indentation for list items at the same level
content/docs/framework/Agile_Documentation.md:88: MD005 Inconsistent indentation for list items at the same level
content/docs/framework/Agile_Documentation.md:89: MD005 Inconsistent indentation for list items at the same level
content/docs/framework/Agile_Documentation.md:90: MD005 Inconsistent indentation for list items at the same level
```

**Comment by jaymzh** [comment]
You're right, according to the spec, [example 266](https://spec.commonmark.org/0.29/#example-266) and [example 267](https://spec.commonmark.org/0.29/#example-267).

According to the spec, it must in fact be aligned with there the text started even if that's odd like:

```
-    thing
     thing
-    mything
```

I'm not entirely certain how this rule is implemented, though I remember some magic on list item parsing. However, I suspect this idea of consistent here is actually not what we want. What we probably want is two separate checks:

1. Did you line up your indentation with the text on the line that started the line-item (correctness)
2. Did you insert multiple spaces between the list-item-marker and the beginning of the text (style)

**Comment by hoijui** [comment]
> @jaymzh The markdown style I'm most familiar with says that
> 
>     * Nested lists should be indented so that the nested list's start token lines up with the parent's text.
> 
>     * There should be a space after a list's start token.
> 
> 
> So two spaces for each level of an unordered list and three spaces for each level of an ordered list.
> 
> Also, for my own reference, when you say that 2 spaces should be preferred, what does that look like with ordered lists? Like this?
> 
> ```gfm
> 1. First item
>   1. Second item
> ```

... and what does it look like if the index gets longer then 1 digit?
does one have to change all items to align with 2 digits,
or just the 2 digit ones?
should we format all lists in a way to expect possilby 2 digits (as items may be added later)?
what if there are more then 2 digits?

**Comment by jamesquilty** [comment]
This bug also affects nested ordered lists, I see MD005 trigger on lines 3 and 4 of the following:

```markdown
1. Unindented ordered item

   1. Intended ordered item
   1. Intended ordered item

1. Unintended ordered item
```

as well as the case where the indented items have the unordered list prefix (as reported in #374).

This is a big enough problem for me to disable MD005 and rely on my Markdown formatter to enforce a consistent indentation... but that's not ideal and it would be great for this bug to be fixed.

> ... and what does it look like if the index gets longer then 1 digit?

 That question is the reason why [MD029](https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md029---ordered-list-item-prefix) defaults to `:one`, i.e. the ordered list marker is _always_ `1.`.

**Comment by hoijui** [comment]
> This bug also affects nested ordered lists, I see MD005 trigger on lines 3 and 4 of the following:
> 
> ```
> 1. Unindented ordered item
> 
>    1. Intended ordered item
>    1. Intended ordered item
> 
> 1. Unintended ordered item
> ```

I do not see any error with this Markdown snippet. Are you using non-default settings?

**Comment by hoijui** [comment]
To work towards a solution ...

As I understand it, the main issue here is, that MD005 looks only at the depth of an item -disregarding which previous levels are ordered or unordered - and then compares its indent to the previously found one of an item at the same depth.

I see two possible solutions:

1. Remember/Record only one indent (e.g. 2 spaces) for unordered lists (after the previous level), and only one indent for ordered lists (e.g. 3), and then require that at each depth for the respective list type (ordered or unordered)
2. Remember/Record the full indent for each combination of list-types, e.g. `OUUOU = 12`, meaning an item in an unordered list at depth 5, with the previous levels or list being ordered, unordered, unordered, ordered, needs to be indented with 12.

What do you say?

**Comment by jamesquilty** [comment]
> I do not see any error with this Markdown snippet. Are you using non-default settings?

 I am using non-default settings, but nothing that should affect the output, the problem is that I didn't provide an example which reproduces the false positive and I didn't mention the version I'm using. Apologies, that was negligent!

The _minimal_ case to reproduce, with output with all rules enabled, is:

```text
$ mdl
# First line in file must be a top level header

- Unindented unordered item
  - Indented unordered item

1. Unindented ordered item

   1. Intended ordered item
   1. Intended ordered item

1. Unintended ordered item
^D
(stdin):6: MD005 Inconsistent indentation for list items at the same level
(stdin):7: MD005 Inconsistent indentation for list items at the same level
(stdin):2: MD007 Unordered list indentation
```

The thing that misled me was that the false positive was raised when the unordered and ordered lists are separated by another list:

```text
$ mdl
# First line in file must be a top level header

- Unindented unordered item
  - Indented unordered item
  - ...
  - Final indented unordered item

Some line of text

- Unindented unordered item
- ...
- Unindented unordered item

## Header at any level

1. Unindented ordered item

   1. Intended ordered item
   1. Intended ordered item

1. Unintended ordered item
^D
(stdin):18: MD005 Inconsistent indentation for list items at the same level
(stdin):19: MD005 Inconsistent indentation for list items at the same level
(stdin):4: MD007 Unordered list indentation
```

The MD007 violation is independent, that rule can be disabled and the MD005 false positives are still observed.

I think that the violations flagged on lines 18 and 19 are invalid.

**Comment by hoijui** [comment]
Ahh I see, thank you!
then it makes sense.. and it is the same problem we others observed, and it fits the summary of the issue in my previous comment.

Any though on my proposed solutions or an alternative one?

**Comment by jamesquilty** [comment]
There's certainly a question over whether [MD005](https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md005---inconsistent-indentation-for-list-items-at-the-same-level) should consider _local consistency_ or _global consistency_ of list indentation. 

The MD005 title, description and examples imply the intended behaviour is _local consistency_ within each indented block, and this is definitely the simpler case to address.

Considering _global consistency_ in MD005, as the comment "MD005 requires consistent indentation across lists" in [MD007](https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md007---unordered-list-indentation) implies, will be _very_ complicated due to the many different cases that will need to be accommodated.

I think that making MD005 consider _local consistency_ only will add value, including being able to simplify the advice in MD007. If _global consistency_ of list indentation is seen to be a good thing then a separate rule should be constructed to address it.

**Relevant:** I use [mdformat](https://github.com/executablebooks/mdformat) which applies different indentation to nested lists within ordered (2 spaces) and unordered (3 spaces) lists. It's not useful for Markdownlint to flag a violation of MD005 when mdformat is being used. Incidentally, mdformat uses `1.` as the ordered list marker for all items.

**Comment by hoijui** [comment]
100% agree, with everything.

**Comment by HariSekhon** [comment]
MD005 is the bane of my README editing life.

I also prefer 2 space indentation and override the default spacing which works fine for unordered lists, but as soon as you add an ordered list, it requires 3 space indentation otherwise it doesn't render properly.

This then flags the ordered and unordered lists as having different indentations.

I've disabled this rule entirely for now on one of my repos because I can't see how to fix this without moving all my unordered lists to 3 space intentions which look ugly.

IMO this rule should be updated to separate indentation handling for ordered vs unordered lists due to the difference in rendering.
